core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -130,45 +148,95 @@ async def step3_confirm_tx(self) -> None:
         if fee < 0:
             self.on_negative_fee()
 
+        total = self.total_in - self.change_out
+        spending = total - self.external_in
+
         # fee > (coin.maxfee per byte * tx size)
         if fee > (self.coin.maxfee_kb / 1000) * (self.weight.get_total() / 4):
             await helpers.confirm_feeoverthreshold(fee, self.coin)
         if self.tx.lock_time > 0:
             await helpers.confirm_nondefault_locktime(self.tx.lock_time)
-        await helpers.confirm_total(self.total_in - self.change_out, fee, self.coin)
+        if total == spending:
+            await helpers.confirm_total(spending, fee, self.coin)
+        else:
+            await helpers.confirm_joint_total(spending, total, self.coin)
 
-    async def step4_serialize_inputs(self) -> None:
+    async def step4_verify_external_inputs(self) -> None:
+        # should come out the same as h_external, checked before continuing
+        h_check = self.create_hash_writer()
+
+        for i in sorted(self.external):
+            progress.advance()
+            txi = await helpers.request_tx_input(self.tx_req, i, self.coin)
+            writers.write_tx_input_check(h_check, txi)
+            prev_amount, script_pubkey = await self.get_prevtx_output(
+                txi.prev_hash, txi.prev_index
+            )
+            if prev_amount != txi.amount:
+                raise wire.DataError("Invalid amount specified")
+
+            if txi.ownership_proof:
+                if not verify_nonownership(
+                    txi.ownership_proof, script_pubkey, b"", self.keychain, self.coin
+                ):
+                    raise wire.DataError("Fake external input")
+            else:
+                verifier = SignatureVerifier(
+                    script_pubkey, txi.script_sig, txi.witness, self.coin
+                )
+
+                if not verifier.check_hash_type(self.get_hash_type()):
```

Comment:
> * it's unclear that `self.get_hash_type()` returns a "constant" value, as opposed to some dynamic property of the current transaction

The fact that it's constant is just a consequence of the fact that we currently only support SIGHASH_ALL. If we supported different hash types, then it could be different for every input.

check_hash_type() is a sort of sanity check, it doesn't serve a security purpose, we could remove it. If someone is using a different hash type then the signature verification will fail. This just gives a clearer error message.